tools/nimgrep.nim

```diff
@@ -223,242 +324,357 @@ proc writeArrow(s: string) =
 proc blockHeader(filename: string, line: int|string, replMode=false) =
   if replMode:
     writeArrow("     ->\n")
-  elif newLine:
+  elif newLine and optFilenames notin options:
     if oneline:
       printBlockFile(filename)
       printBlockLineN(":" & $line & ":")
     else:
       printBlockLineN($line.`$`.align(alignment) & ":")
     stdout.write("\n")
 
-proc lineHeader(filename: string, line: int|string, isMatch: bool) =
+proc newLn(curCol: var Column) =
+  stdout.write("\n")
+  curCol.file = 0
+  curcol.terminal = 0
+
+proc lineHeader(filename: string, line: int|string, isMatch: bool,
+                curCol: var Column) =
   let lineSym =
     if isMatch: $line & ":"
     else: $line & " "
-  if not newLine:
+  if not newLine and optFilenames notin options:
     if oneline:
       printFile(filename)
       printLineN(":" & lineSym, isMatch)
+      curcol.terminal += filename.len + 1 + lineSym.len
     else:
       printLineN(lineSym.align(alignment+1), isMatch)
-    stdout.write(" ")
+      curcol.terminal += lineSym.align(alignment+1).len
+    stdout.write(" "); curCol.terminal += 1
 
-proc printMatch(fileName: string, mi: MatchInfo) =
-  let lines = mi.match.splitLines()
-  for i, l in lines:
+proc printMatch(fileName: string, mi: MatchInfo, curCol: var Column) =
+  let sLines = mi.match.splitLines()
+  for i, l in sLines:
     if i > 0:
-      lineHeader(filename, mi.lineBeg + i, isMatch = true)
-    writeColored(l)
-    if i < lines.len - 1:
-      stdout.write("\n")
+      lineHeader(filename, mi.lineBeg + i, isMatch = true, curCol)
+    if curCol.terminal < limitChar:
+      writeColored(l)
+    else:
+      curCol.overflowMatches += 1
+    if i < sLines.len - 1:
+      newLn(curCol)
+  curCol.terminal += mi.match.len
+  curCol.file += mi.match.len
+
+const matchPaddingFromRight = 10
+let ellipsis = "..."
+
+proc reserveChars(mi: MatchInfo): int =
+  if optLimitChars in options or optFit in options:
+    let patternChars = afterPattern(mi.match, 0) + 1
+    result = patternChars + ellipsis.len + matchPaddingFromRight
+  else:
+    result = 0
+
+proc printRaw(c: char, curCol: var Column, allowTabs = true) =
+  # print taking into account tabs and optOnlyAscii
+  if c == '\t':
+    if allowTabs:
+      let spaces = 8 - (curCol.file mod 8)
+      curCol.file += spaces
+      curCol.terminal += spaces
+      if optOnlyAscii:
+        printSpecial " "
+        stdout.write " ".repeat(spaces-1)
+      else:
+        stdout.write " ".repeat(spaces)
+    else:
+      curCol.file += 1
+      curCol.terminal += 1
+      if optOnlyAscii:
+        printSpecial " "
+      else:
+        stdout.write " "
+  elif not optOnlyAscii or (0x20 <= int(c) and int(c) <= 0x7e):
+    stdout.write c
+    curCol.file += 1
+    curCol.terminal += 1
+  else:  # substitute characters that are not ACSII Latin
+    let substitute =
```

Comment:
I don't understand this, it results in hard to understand ascii being written; why not simply use `repr` ?

eg look at output of:
`for i in 0..<255: echo i, ":", repr(i.char)`

Reply:
it's more concise, just one char => also more space will be available in --cols:auto. repr(i.char) contains 1,2, or 3 chars.

I think there is no actually demand to interpret this, rather just highlight the presence of a non-printable symbol.

BTW vim also print such characters in this form, though it adds ^, so '\12' in it is represented as ^L (in blue).

Reply:
> it's more concise, just one char

but you're collapsing a range of chars to just `!`, that's not good; if I want to print binary as ascii, I want to be able to interpret it, otherwise I'd just ask to skip.

see also how less renders binary data: (consecutive char's):
```
^@^A^B^C^D^E^F^G^H
^K^L^M^N^O^P^Q^R^S^T^U^V^W^X^Y^Z\^]^^^_ !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~^?<80><81><82><83><84><85><86><87><88><89><8A><8B><8C><8D><8E><8F><90><91><92><93><94><95><96><97><98><99><9A><9B><9C><9D><9E><9F>
<A0><A1><A2><A3><A4><A5><A6><A7><A8><A9><AA><AB><AC><AD><AE><AF><B0><B1><B2><B3><B4><B5><B6><B7><B8><B9><BA><BB><BC><BD><BE><BF><C0><C1><C2><C3><C4><C5><C6><C7><C8><C9><CA><CB><CC><CD><CE><CF><D0><D1><D2><D3><D4><D5><D6><D7><D8><D9><DA><DB><DC><DD><DE><DF><E0><E1>
<E2><E3><E4><E5><E6><E7><E8><E9><EA><EB><EC><ED><EE><EF><F0><F1><F2><F3><F4><F5><F6><F7><F8><F9><FA><FB><FC><FD><FE><FF>^@^A^B^
```
more readable/interpretable than your current scheme.

 and shorter than `repr`


how about following:
```nim
import sugar, strutils
proc render(result: var string, c: char) =
  result.add '\\' & c.BiggestUInt.toHex(2)
doAssert "".dup(render(243.char)) == r"\F3"
```
which is 1 char shorter that repr (`\F3` instead of `\243`)

Reply:
OK, I chose representation like ^@ and 'FF for 0xFF. I hope it doesn't seem too strange.